Corrected the numbers of iterations for multi-step time integrators

joshuahansel opened the pull request:
I added a test for each time integrator, which may have been a bit excessive, but since the iteration counting isn't confined to the base, it seemed the safest route. On my machine, these tests took 4 seconds total. At the very least, we should have tests for one single-stage time integrator and all of the time integrators that override `solve()`.

Closes #11444

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

joshuahansel committed:
Corrected the numbers of iterations for multi-step time integrators

Closes #11444

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/187870/) on 6f9e697 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11472/site)

This comment will be updated on new commits.

joshuahansel commented:
I went ahead and set `max_parallel = 1`.

joshuahansel commented:
@andrsd Could you review this?

joshuahansel commented on framework/src/timeintegrators/TimeIntegrator.C:
```diff
@@ -36,12 +40,36 @@ TimeIntegrator::TimeIntegrator(const InputParameters & parameters)
+}
+
+unsigned int
+TimeIntegrator::getNumLinearIterationsLastPetscSolve() const
```
The reason I put `Petsc` in the name is because it only applies if PETSc was used for the nonlinear solver. `ActuallyExplicitEuler`, for example, uses a libMesh solver, and so if someone called `getNumLinearIterationsLastSolve`, they'd get back an incorrect number of linear iterations. Do you still think I should remove it? If so, I can at least put a caution in the doxygen description.

andrsd replied:
I was thinking for the future where this function would give back the number of linear iterations no matter what the solver was...

Could `ActuallyExplicitEuler` override this function and return the right number?

joshuahansel replied:
This should be possible, but keep in mind that this function doesn't even have `public` access; it is only used internally by `TimeIntegrator` classes. So what would happen here in `ActuallyExplicitEuler` would be this:
```
getNumLinearIterationsLastSolve() {return _n_linear_iterations;}
```
but it wouldn't be used anywhere because `_n_linear_iterations` is already set. Or we'd change
```
_n_linear_iterations = num_its_and_final_tol.first;
```
to
```
_n_linear_iterations_last_solve = num_its_and_final_tol.first;
```
and then do
```
_n_linear_iterations = getNumLinearIterationsLastSolve();
```
where
```
getNumLinearIterationsLastSolve() {return _n_linear_iterations_last_solve;}
```
What would you prefer to do? I'll need to make changes in any case because I need to add requirements, etc.

andrsd replied:
Just for the record, we decided to drop the `Petsc` part from the method name, because we are able to call the solver independent interface in libmesh which is `NonlinearSolver<T>`.

permcody commented:
@joshuahansel - Would you please add the "requirement, design, and issue" tags to all of your new tests? Thanks!

Take a look at the bottom of this page:
http://mooseframework.org/moose/utilities/MooseDocs/generate.html

joshuahansel committed:
Made modifications to MooseDocs page

Some grammar issues were fixed, and an example for the
requirements, design, and issues for new tests was added.

joshuahansel commented:
I also made some changes to the MooseDocs page that Cody pointed me to (grammar and an example for the the tests requirements stuff).
